File: AbstractDataType/AbstractDataType/ArrayList.cs
public class ArrayList
{
    private int[] buffer;
    private int count;
    
    public int Count
    {
        get { return count; } 
    }

    public int this[int i]
    {   set 
        { if (i > count || i < 0)
            {
                
            }else 
            {
                buffer[i] = value;
            }
        }
        get 
        {if (i >= count || i < 0)
            {
                return -1;
            }
            else { return buffer[i]; }
        }
    }

    public ArrayList() 
    {
        buffer = null;
        count = 0;
    }
    
    public void Insert(int num, int index)
    {
        if (index < 0 || index > count)
        {

        }
        else
        {
            Expand();
            for (int i = count; i > index; i--)
            {
                buffer[i] = buffer[i - 1];
            }
            buffer[index] = num;
            count++;
        }
    }

    public void Delete(int index)
    {
        if (index < 0 || index >= count || count==0)
        {
            
        }
        else
        {
            if (count == 1) 
            {
                Clear();
            }
            else 
            {
            for (int i = index; i < count - 1; i++) { buffer[i] = buffer[i + 1]; }
            buffer[count - 1] = 0;
            count--;
            }      
        }
    }
    public void Add(int num) 
    {
        Expand();
        buffer[count] = num;
        count++;
    }

    public void Clear() 
    {
        count = 0;
        buffer = null;
    }

    private void Expand()
    {
        if (buffer != null)
        {
            int[] buffer = new int[count * 2];
            for (int i = 0; i < count; i++)
            {
                buffer[i] = this.buffer[i];
            }
            this.buffer = buffer;
        }
        else buffer = new int[1];
    }
    
    public void Print()
    {
        if (buffer != null)
        {
            for (int i = 0; i < count; i++) Console.WriteLine(buffer[i]);
        }
    }
}

File: Base++/ArrayList.cpp
#include "pch.h"

class ArrayList : public BaseList {
private:
	int* buffer;
	int buffer_length;
	BaseList* EmptyClone() override {
		return new ArrayList();
	}
	void Expand() {
		if (count == buffer_length) {
			if (buffer_length == 0) {
				buffer = new int[1];
				buffer_length = 1;
			}
			else {
				buffer_length = buffer_length * 2;
				int* temp_buffer = new int[buffer_length];
				for (int i = 0; i < count; i++) {
					temp_buffer[i] = buffer[i];
				}
				delete[] buffer;
				buffer = temp_buffer;
			}
		}
	}
public:
	ArrayList() {
		buffer = nullptr;
		buffer_length = 0;
		count = 0;
	}

	~ArrayList() {
		delete[] buffer;
	}

	void Add(int number) override {
		Expand();
		buffer[count] = number;
		count++;
	}

	void Insert(int number, int index) override {
		if (index<0 || index>count) {
			std::cout << "ARRAY CHLENIX" << '\t';
			std::cout << "count = " << count << " index = " << index << '\n';
		}
		else {
			Expand();
			for (int i = count; i > index; i--) {
				buffer[i] = buffer[i - 1];
			}
			buffer[index] = number;
			count++;
		}
	}

	void Delete(int index) override {
		if (index < 0 || index >= count) { std::cout << "NAM PIZDA" << '\n'; }
		else {
			if (count == 1) {
				Clear();
			}
			else {
				for (int i = index; i < count - 1; i++) { buffer[i] = buffer[i + 1]; }
				buffer[count - 1] = 0;
				count--;
			}
		}
	}

	/*void Print() override {
		for (int i = 0; i < Count(); i++) {
			std::cout << buffer[i] << "\t";
		}
		std::cout << "\n";
	}*/

	void Clear() override {
		delete[] buffer;
		count = 0;
		buffer_length = 0;
		buffer = nullptr;
	}

	int& operator[](int i) override {
		static int mock = 0;
		if (i < 0 || i >= count) return mock;
		else return buffer[i];
	}
};